src/main/java/io/cdap/plugin/batch/source/BSONObjectToRecordTransformer.java

```diff
@@ -0,0 +1,170 @@
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.bson.BSONObject;
+import org.bson.BsonUndefined;
+import org.bson.types.BasicBSONList;
+import org.bson.types.Decimal128;
+import org.bson.types.ObjectId;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Transforms {@link BSONObject} to {@link StructuredRecord}.
+ */
+public class BSONObjectToRecordTransformer {
+
+  private final Schema schema;
+
+  public BSONObjectToRecordTransformer(Schema schema) {
+    this.schema = schema;
+  }
+
+  /**
+   * Transforms given {@link BSONObject} to {@link StructuredRecord}.
+   *
+   * @param bsonObject BSON object to be transformed.
+   * @return {@link StructuredRecord} that corresponds to the given {@link BSONObject}.
+   */
+  public StructuredRecord transform(BSONObject bsonObject) {
+    return convertRecord(bsonObject, schema);
+  }
+
+  private StructuredRecord convertRecord(Object object, Schema schema) {
+    BSONObject bsonObject = (BSONObject) object;
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = Objects.requireNonNull(schema.getFields(), "Schema fields cannot be empty");
+    for (Schema.Field field : fields) {
+      Object value = extractValue(bsonObject.get(field.getName()), field.getSchema());
+      Schema.LogicalType fieldLogicalType = field.getSchema().isNullable()
+        ? field.getSchema().getNonNullable().getLogicalType()
+        : field.getSchema().getLogicalType();
+
+      if (fieldLogicalType == null) {
+        // Set values of simple types
+        builder.set(field.getName(), value);
+        continue;
+      }
+
+      // Set values of logical types properly
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          builder.setTimestamp(field.getName(), (ZonedDateTime) value);
+          break;
+        case DECIMAL:
+          builder.setDecimal(field.getName(), (BigDecimal) value);
+          break;
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
+      }
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(Object object, final Schema schema) {
+    if (object == null || object instanceof BsonUndefined) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.isNullable()
+      ? schema.getNonNullable().getLogicalType()
+      : schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          Date date = (Date) object;
```

Comment:
If the document doesn't match the schema, all of these casts will fail and the logs will contain a mostly unhelpful message. We need to make sure the failure includes the name of the field that had the issue. Before we cast anything, we need to check it's the expected type. If not, we need to throw an exception with a message like:

Document field '%s' is expected to be of type '%s', but found a '%s'.

Reply:
Created the following method:
```
  private void ensureTypeValid(String fieldName, Object value, Class... expectedTypes) {
    for (Class expectedType : expectedTypes) {
      if (expectedType.isInstance(value)) {
        return;
      }
    }
	
    String expectedTypeNames = Stream.of(expectedTypes)
      .map(Class::getName)
      .collect(Collectors.joining(", "));
    throw new UnexpectedFormatException(
      String.format("Document field '%s' is expected to be of type '%s', but found a '%s'.", fieldName,
                    expectedTypeNames, value.getClass().getName()));
  }
```

along with [unit tests](https://github.com/data-integrations/mongodb-plugins/pull/3/files#diff-19f4357d1cb90adb8abc786cce615966R209).